File: omnibus-ios/omnibus/Reader/ReadStatusAuto.swift
//  ReadStatusAuto.swift
//  Automatic read-status transitions driven by the readers and the audio
//  player: starting an unread book marks it reading, and reaching its end
//  marks it finished. The native port of the web `read_status_auto`.

import Foundation

/// One book's automatic transitions, owned by the reader that has it open.
///
/// A decision is never made against a guessed status — writing `reading` over
/// an unfetched `finished` would be a downgrade — so an unknown status keeps
/// every transition inert. It is retried on later observations rather than
/// settled once at open, because the opening fetch is exactly what a book
/// opened offline loses. Writes take the same outbox path as the detail
/// screen's manual chip (`UserDataService.setReadStatus`), so they queue
/// offline like any other content-state write.
@MainActor
final class ReadStatusAuto {
    /// `nil` until a fetch lands — the tracker knows nothing, and every
    /// transition stays inert until it does.
    private var status: ReadStatus?
    private var atEnd = false
    private let fetch: () async -> ReadStatus?
    private let write: (ReadStatus) async -> Void
    private let isOnline: () -> Bool

    /// The production wiring for one book, over `UserDataService`.
    convenience init(uuid: String) {
        self.init(
            fetch: { await UserDataService.storedReadStatus(uuid: uuid) },
            write: { await UserDataService.setReadStatus(uuid: uuid, status: $0) },
            isOnline: { Connectivity.shared.isOnline }
        )
    }

    init(
        fetch: @escaping () async -> ReadStatus?,
        write: @escaping (ReadStatus) async -> Void,
        isOnline: @escaping () -> Bool = { true }
    ) {
        self.fetch = fetch
        self.write = write
        self.isOnline = isOnline
    }

    /// The status to write for a book whose stored state is `current`,
    /// observed either at the reader's end position (`atEnd`) or merely open.
    /// `nil` means no write: opening never downgrades a finished book back to
    /// reading, and re-reaching the end of a finished book is a no-op.
    nonisolated static func transition(current: ReadStatus, atEnd: Bool) -> ReadStatus? {
        if atEnd {
            return current == .finished ? nil : .finished
        }
        return current == .unread ? .reading : nil
    }

    /// Fetch the stored status and apply the open transition. Call once per
    /// open, off the reader's critical path — the fetch may cost a round trip.
    func bookOpened() async {
        await fetchIfUnknown()
        await applyIfNeeded()
    }

    /// Tell the tracker where the reader is. Cheap on every relocate or page
    /// turn; an end observed before the fetch lands is applied when it does.
    ///
    /// An unknown status is retried here rather than latched at open. The
    /// opening fetch is the one every book opened offline loses — the replica
    /// has no row to fall back on for a book nobody has marked — and giving up
    /// on it there is what left a book read cover to cover on a plane with no
    /// status at all (#2289).
    func positionChanged(atEnd: Bool) async {
        self.atEnd = atEnd
        // Reaching the end retries whatever the connection is doing: unlike
        // the open transition it cannot downgrade anything, and dropping it
        // loses the strongest completion signal the app gets. The audio player
        // relies on that — it finishes books it was never asked to open. An
        // ordinary relocate only retries when there is a server to ask, so
        // reading offline costs nothing until the device is back.
        if atEnd || isOnline() {
            await fetchIfUnknown()
        }
        await applyIfNeeded()
    }

    /// Learn the stored status, unless this tracker already knows it.
    ///
    /// A status this tracker already holds is kept rather than refetched: it
    /// is either the one this fetch would return, or the newer one this
    /// tracker just wrote. Overwriting the second with a response that
    /// predates it re-runs a transition that has already happened, and writes
    /// `finished` twice.
    ///
    /// That is why the check is repeated **after** the fetch and not only
    /// before it. `fetch` suspends, and every relocate drives this — so a
    /// second observation can settle the status, and write it, while the
    /// first request is still in flight. Assigning unconditionally on the way
    /// out replays that transition, or drops a settled status back to unknown
    /// when the late answer is a failure. Two concurrent fetches can still
    /// both go out; that costs a redundant GET and nothing else.
    private func fetchIfUnknown() async {
        guard status == nil else { return }
        let fetched = await fetch()
        guard status == nil else { return }
        status = fetched
    }

    private func applyIfNeeded() async {
        guard let current = status,
              let next = Self.transition(current: current, atEnd: atEnd)
        else { return }
        // Optimistic: move the tracked status before the write, so the next
        // observation settles on no-transition instead of repeating a write
        // that is still queued.
        status = next
        await write(next)
    }
}

File: frontend/src/read_status_auto.rs
//! Automatic read-status transitions driven by the readers and the audio
//! players: starting an `Unread` book marks it `Reading`, and reaching its
//! end marks it `Finished`. Pure decision logic, the hook the EPUB reader and
//! the CBZ pager mount, and the one-shot [`apply_auto_read_status`] the
//! players call from their media-event callbacks.

use dioxus::prelude::*;
use omnibus_shared::{ReadStatus, SetReadStatus};

use crate::data;

#[cfg(test)]
mod tests;

/// The status to write for a book whose stored state is `current`, observed
/// either at the reader's end position (`at_end`) or merely open. `None`
/// means no write: opening never downgrades a `Finished` book back to
/// `Reading`, and re-reaching the end of a `Finished` book is a no-op.
pub fn auto_transition(current: ReadStatus, at_end: bool) -> Option<ReadStatus> {
    if at_end {
        (current != ReadStatus::Finished).then_some(ReadStatus::Finished)
    } else {
        (current == ReadStatus::Unread).then_some(ReadStatus::Reading)
    }
}

/// Apply [`auto_transition`] once, outside a component.
///
/// The readers observe `at_end` as reactive state and drive it through
/// [`use_auto_read_status`]. The audio players have no such signal — they are
/// driven by media events fired from JS — so they call this directly: `false`
/// on the first play of a book, `true` when the last part reaches its end.
///
/// Best-effort like the readers' writes, and for the same reason a failed
/// *fetch* decides nothing: writing `Reading` over an unfetched `Finished`
/// would be a downgrade. The next event retries.
pub async fn apply_auto_read_status(server_url: &str, uuid: &str, at_end: bool) {
    let Ok(stored) = data::get_read_status(server_url, uuid).await else {
        return;
    };
    let Some(next) = auto_transition(stored.map(|r| r.status).unwrap_or_default(), at_end) else {
        return;
    };
    let _ = data::set_read_status(
        server_url,
        SetReadStatus {
            book_uuid: uuid.to_string(),
            status: next,
        },
    )
    .await;
}

/// Drive [`auto_transition`] for the book open in a reader: fetch the stored
/// status once per book, then re-apply whenever the status or the reader's
/// `at_end` position changes. Writes are best-effort like the readers'
/// progress saves — a failed write must not interrupt reading, and the next
/// open retries. Call unconditionally from the page (rule 07); on SSR the
/// effects never fire and this is a no-op.
pub fn use_auto_read_status(uuid: String, server_url: String, at_end: Memo<bool>) {
    // The fetched status, keyed by the uuid the fetch answered for: the
    // write effect reads the pair together, so it can never decide against
    // — or write to — a book the fetch didn't describe. `None` until the
    // fetch lands, and stays `None` when it fails, so a transition is never
    // decided against a guessed status (writing `Reading` over an unfetched
    // `Finished` would be a downgrade).
    let mut status: Signal<Option<(String, ReadStatus)>> = use_signal(|| None);
    // Monotonic fetch guard (`load_seq` in `pages/book_detail/read_status.rs`
    // is the model): a route-param book swap reuses the mounted page, so a
    // slow response for the previous book must not land as the current one's.
    let mut load_seq = use_signal(|| 0u64);

    {
        let server_url = server_url.clone();
        use_effect(use_reactive!(|uuid| {
            let my_load = *load_seq.peek() + 1;
            load_seq.set(my_load);
            status.set(None);
            let server_url = server_url.clone();
            spawn(async move {
                if let Ok(stored) = data::get_read_status(&server_url, &uuid).await {
                    if *load_seq.peek() == my_load {
                        let fetched = stored.map(|r| r.status).unwrap_or_default();
                        status.set(Some((uuid, fetched)));
                    }
                }
            });
        }));
    }

    use_effect(move || {
        let Some((book_uuid, current)) = status() else {
            return;
        };
        let Some(next) = auto_transition(current, at_end()) else {
            return;
        };
        // Optimistic: move the signal first so the re-run settles on `None`
        // instead of repeating the write while the POST is in flight.
        status.set(Some((book_uuid.clone(), next)));
        let server_url = server_url.clone();
        let update = SetReadStatus {
            book_uuid,
            status: next,
        };
        spawn(async move {
            let _ = data::set_read_status(&server_url, update).await;
        });
    });
}
